lib/charms/mongodb_libs/v0/mongodb_provider.py

```diff
@@ -159,25 +180,37 @@ def _on_legacy_relation_created(self, event):
         """
         logger.warning("DEPRECATION WARNING - `mongodb` interface is a legacy interface.")
 
-        # TODO, future PR check if there are any new relations that are related to this charm and will be effected
-        # by loss of password. If so go into blocked state.
-
-        # TODO, future PR check if already running without auth
-        try:
-            self._stop_mongod_service()
-            self._update_mongod_service(auth=False)
-            self._start_mongod_service()
-        except systemd.SystemdError:
-            self.charm.unit.status = BlockedStatus("couldn't restart MongoDB")
+        # legacy relations turn off authentication, therefore disabling authentication for current
+        # users (which connect over the new relation interface). If current users exist that use
+        # auth it is necessary to not proceed and go into blocked state.
+        relation_users = self._get_users_from_relations(departed_relation_id=None)
+        if len(relation_users) > 0:
+            self.charm.unit.status = BlockedStatus("cannot have both legacy and new relations")
+            logger.error(
+                "Creating legacy relation would turn off auth effecting the new relations: %s",
+                relation_users,
+            )
             return
 
+        # If auth is already disabled its likely it has a connection with another legacy relation
+        # user. Shutting down and restarting mongod would lead to downtime for the other legacy
+        # relation user and hence shouldn't be done. Not to mention there is no need to disable
+        # auth if it is already disabled.
+        if self._auth_enabled:
+            try:
+                logger.debug("Disabling authentication.")
+                self._stop_mongod_service()
+                self._update_mongod_service(auth=False)
+                self._start_mongod_service()
+            except systemd.SystemdError:
+                self.charm.unit.status = BlockedStatus("couldn't restart MongoDB")
+                return
+
```

Comment:
Could you generalise `oversee_users` and call that here?

Reply:
For now I am going to keep these separate, I think it will make things simpler when it comes time to remove everything related to legacy relations. Also `oversee_users` has a bunch of new relation code that would be challenging/annoying for me to generalise